File: java/javasearch/lib/src/main/java/javasearch/Searcher.java
/*******************************************************************************
Searcher

Class to perform the file search

@author Cary Clark &lt;clarkcb@gmail.com&gt;
@version $Rev$
@copyright Cary Clark 2012
*******************************************************************************/

package javasearch;

import javafind.*;
import org.apache.commons.io.LineIterator;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static javafind.Logger.log;

public class Searcher {

    final private SearchSettings settings;
    final private Finder finder;
    private Charset charset;

    public Searcher(final SearchSettings settings) {
        this.settings = settings;
        this.finder = new Finder(settings);
    }

    public SearchSettings getSettings() {
        return this.settings;
    }

    final void validateSettings() throws SearchException {
        try {
            this.finder.validateSettings();
        } catch (FindException e) {
            throw new SearchException(e.getMessage());
        }
        if (settings.getSearchPatterns().isEmpty()) {
            throw new SearchException("No search patterns defined");
        }
        if (settings.getLinesAfter() < 0) {
            throw new SearchException("Invalid linesafter");
        }
        if (settings.getLinesBefore() < 0) {
            throw new SearchException("Invalid linesbefore");
        }
        if (settings.getMaxLineLength() < 0) {
            throw new SearchException("Invalid maxlinelength");
        }
        try {
            charset = Charset.forName(settings.getTextFileEncoding());
        } catch (IllegalArgumentException e) {
            throw new SearchException("Invalid encoding provided");
        }
    }

    private boolean anyMatchesAnyPattern(final List<String> sList,
                                         final Set<Pattern> patternSet) {
        return sList.stream().anyMatch(s -> matchesAnyPattern(s, patternSet));
    }

    private boolean matchesAnyPattern(final String s,
                                      final Set<Pattern> patternSet) {
        return null != s && patternSet.stream().anyMatch(p -> p.matcher(s).find());
    }

    public final List<SearchResult> search() throws SearchException {
        List<FileResult> fileResults;
        try {
            fileResults = finder.find();
        } catch (FindException e) {
            throw new SearchException(e.getMessage());
        }

        if (settings.getVerbose()) {
            List<String> dirResults = fileResults.stream()
                    .map(fr -> fr.path().getParent())
                    .map(p -> p == null ? "." : p.toString())
                    .distinct()
                    .sorted(String::compareTo)
                    .toList();
            log(String.format("\nDirectories to be searched (%d):",
                    dirResults.size()));
            for (String d : dirResults) {
                log(d);
            }
            log(String.format("\n\nFiles to be searched (%d):",
                    fileResults.size()));
            for (FileResult fr : fileResults) {
                log(fr.toString());
            }
            log("");
        }

        List<SearchResult> searchResults = searchFiles(fileResults);

        if (settings.getVerbose()) {
            log("\nFile search complete.\n");
        }

        sortSearchResults(searchResults);
        return searchResults;
    }

    private List<SearchResult> searchFiles(List<FileResult> fileResults) {
        List<SearchResult> results = new ArrayList<>();

        int offset = 0;
        int batchSize = 1000;

        if (fileResults.size() > batchSize) {
            do {
                int toIndex = Math.min(offset + batchSize, fileResults.size());
                List<FileResult> nextBatch = fileResults.subList(offset, toIndex);
                results.addAll(batchSearchFiles(nextBatch));
                offset += batchSize;
            } while (offset < fileResults.size());
        } else {
            for (FileResult fr : fileResults) {
                results.addAll(searchFile(fr));
            }
        }

        return results;
    }

    private List<SearchResult> batchSearchFiles(List<FileResult> fileResults) {
        //System.out.println("Next batch: " + fileResults.size() + " searchFiles");
        List<SearchResult> results = new ArrayList<>();

        List<CompletableFuture<List<SearchResult>>> futures = new ArrayList<>(fileResults.size());
        for (FileResult fr : fileResults) {
            futures.add(CompletableFuture.supplyAsync(() -> searchFile(fr)));
        }
        for (CompletableFuture<List<SearchResult>> future : futures) {
            try {
                results.addAll(future.get());
            } catch (InterruptedException | ExecutionException e) {
                log(e.toString());
            }
        }
        return results;
    }

    public final List<SearchResult> searchFile(final FileResult fr) {
        FileType fileType = fr.fileType();
        if (fileType == FileType.CODE || fileType == FileType.TEXT || fileType == FileType.XML) {
            return searchTextFile(fr);
        } else if (fileType == FileType.BINARY) {
            return searchBinaryFile(fr);
        }
        return new ArrayList<>();
    }

    private List<SearchResult> searchTextFile(final FileResult fr) {
        if (settings.getVerbose()) {
            log(String.format("Searching text file %s", fr.toString()));
        }
        if (settings.getMultiLineSearch()) {
            return searchTextFileContents(fr);
        } else {
            return searchTextFileLines(fr);
        }
    }

    private List<SearchResult> searchTextFileContents(final FileResult fr) {
        List<SearchResult> results = new ArrayList<>();
        try {
            String contents = FileUtil.getFileContents(fr.path(), charset);
            results.addAll(searchMultiLineString(contents));
            for (SearchResult r : results) {
                r.setFileResult(fr);
            }
        } catch (NoSuchElementException e) {
            log(e.toString() + ": " + fr.toString());
        } catch (IllegalStateException | IOException e) {
            log(e.toString());
        }
        return results;
    }

    private List<Number> getNewLineIndices(final String s) {
        List<Number> newlineIndices = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '\n') {
                newlineIndices.add(i);
            }
        }
        return newlineIndices;
    }

    private List<Number> getStartLineIndicesFromNewLineIndices(final List<Number> newLineIndices) {
        List<Number> startLineIndices = new ArrayList<>();
        startLineIndices.add(0);
        startLineIndices.addAll(newLineIndices.stream()
                .map(i -> i.longValue() + 1).toList());
        return startLineIndices;
    }

    private List<Number> getEndLineIndicesFromNewLineIndices(final String s,
                                                             final List<Number> newLineIndices) {
        List<Number> endLineIndices = new ArrayList<>(newLineIndices);
        endLineIndices.add(s.length() - 1);
        return endLineIndices;
    }

    public final List<SearchResult> searchMultiLineString(final String s) {
        List<SearchResult> results = new ArrayList<>();
        for (Pattern p : settings.getSearchPatterns()) {
            results.addAll(searchMultiLineStringForPattern(s, p));
        }
        return results;
    }

    private List<String> getLinesFromMultiLineString(final String s,
                                                     final List<Number> startIndices,
                                                     final List<Number> endIndices) {
        List<String> lines = new ArrayList<>();
        for (int i = 0; i < startIndices.size(); i++) {
            lines.add(s.substring(startIndices.get(i).intValue(),
                    endIndices.get(i).intValue()));
        }
        return lines;
    }

    private List<SearchResult> searchMultiLineStringForPattern(final String s,
                                                               final Pattern p) {
        Map<Pattern, Integer> patternMatches = new HashMap<>();
        List<SearchResult> results = new ArrayList<>();
        List<Number> newLineIndices = getNewLineIndices(s);
        List<Number> startLineIndices = getStartLineIndicesFromNewLineIndices(newLineIndices);
        List<Number> endLineIndices = getEndLineIndicesFromNewLineIndices(s,
                newLineIndices);
        final int linesBeforeCount = settings.getLinesBefore();
        final int linesAfterCount = settings.getLinesAfter();
        final int sLength = s.length();
        Matcher m = p.matcher(s);
        boolean found = m.find();
        while (found) {
            if (settings.getFirstMatch() && patternMatches.containsKey(p)) {
                found = false;
            } else {
                // get the start line indices before the match index to get the current line number
                List<Number> beforeStartIndices = startLineIndices.stream()
                        .filter(i -> i.intValue() <= m.start())
                        .collect(Collectors.toList());
                int lineNum = beforeStartIndices.size();
                int nextEndLineIndex = endLineIndices.get(beforeStartIndices.size() - 1).intValue();
                if (nextEndLineIndex == sLength - 1) {
                    nextEndLineIndex++;
                }
                int endLineIndex = nextEndLineIndex;
                int startLineIndex = beforeStartIndices
                        .remove(beforeStartIndices.size() - 1).intValue();
                String line = s.substring(startLineIndex, endLineIndex);

                List<String> linesBefore;
                if (linesBeforeCount > 0) {
                    List<Number> beforeEndIndices = endLineIndices.stream()
                            .filter(i -> i.intValue() < m.start())
                            .collect(Collectors.toList());
                    List<Number> linesBeforeStartIndices = ListUtil
                            .takeRight(beforeStartIndices, linesBeforeCount);
                    List<Number> linesBeforeEndIndices = ListUtil
                            .takeRight(beforeEndIndices, linesBeforeCount);
                    linesBefore = getLinesFromMultiLineString(s,
                            linesBeforeStartIndices, linesBeforeEndIndices);
                } else {
                    linesBefore = new ArrayList<>();
                }

                List<String> linesAfter;
                if (linesAfterCount > 0) {
                    List<Number> afterStartIndices = startLineIndices.stream()
                            .filter(i -> i.intValue() > m.start())
                            .limit(linesAfterCount)
                            .collect(Collectors.toList());
                    List<Number> afterEndIndices = endLineIndices.stream()
                            .filter(i -> i.intValue() > endLineIndex)
                            .limit(linesAfterCount)
                            .collect(Collectors.toList());
                    linesAfter = getLinesFromMultiLineString(s,
                            afterStartIndices, afterEndIndices);
                } else {
                    linesAfter = new ArrayList<>();
                }

                if ((linesBefore.isEmpty() || linesBeforeMatch(linesBefore))
                        &&
                        (linesAfter.isEmpty() || linesAfterMatch(linesAfter))) {
                    SearchResult searchResult = new SearchResult(
                            p,
                            null,
                            lineNum,
                            m.start() - startLineIndex + 1,
                            m.end() - startLineIndex + 1,
                            line,
                            linesBefore,
                            linesAfter);
                    results.add(searchResult);
                    patternMatches.put(p, 1);
                }
                found = m.find(m.end());
            }
        }
        return results;
    }

    private List<SearchResult> searchTextFileLines(final FileResult fr) {
        LineIterator it = null;
        List<SearchResult> results = new ArrayList<>();
        try {
            it = FileUtil.getFileLineIterator(fr.path(), settings.getTextFileEncoding());
            results.addAll(searchStringIterator(it));
            for (SearchResult r : results) {
                r.setFileResult(fr);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (it != null) {
                try {
                    it.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return results;
    }

    private boolean linesMatch(final List<String> lines,
                               final Set<Pattern> inPatterns,
                               final Set<Pattern> outPatterns) {
        return ((inPatterns.isEmpty() || anyMatchesAnyPattern(lines, inPatterns))
                &&
                (outPatterns.isEmpty() || !anyMatchesAnyPattern(lines, outPatterns)));
    }

    private boolean linesBeforeMatch(final List<String> linesBefore) {
        return linesMatch(linesBefore, settings.getInLinesBeforePatterns(),
                settings.getOutLinesBeforePatterns());
    }

    private boolean linesAfterMatch(final List<String> linesAfter) {
        return linesMatch(linesAfter, settings.getInLinesAfterPatterns(),
                settings.getOutLinesAfterPatterns());
    }

    private boolean linesAfterToOrUntilMatch(final Iterator<String> it, List<String> linesAfter) {
        Set<Pattern> linesAfterPatterns;
        if (settings.hasLinesAfterToPatterns()) {
            linesAfterPatterns = settings.getLinesAfterToPatterns();
        } else if (settings.hasLinesAfterUntilPatterns()) {
            linesAfterPatterns = settings.getLinesAfterUntilPatterns();
        } else { // should never get here
            linesAfterPatterns = new HashSet<>();
        }
        boolean linesAfterMatch = anyMatchesAnyPattern(linesAfter, linesAfterPatterns);
        while (!linesAfterMatch && it.hasNext()) {
            String nextLine = it.next();
            linesAfter.add(nextLine);
            linesAfterMatch = matchesAnyPattern(nextLine, linesAfterPatterns);
        }
        if (linesAfterMatch) {
            if (settings.hasLinesAfterUntilPatterns()) {
                linesAfter.remove(linesAfter.size() - 1);
            }
        }
        return linesAfterMatch;
    }

    public final List<SearchResult> searchStringIterator(final Iterator<String> it) {
        int lineNum = 0;
        String line;
        final int linesBeforeCount = settings.getLinesBefore();
        final int linesAfterCount = settings.getLinesAfter();
        List<String> linesBefore = new ArrayList<>();
        List<String> linesAfter = new ArrayList<>();
        Set<Pattern> matchedPatterns = new HashSet<>();
        List<SearchResult> results = new ArrayList<>();
        while (true) {
            lineNum++;
            if (!linesAfter.isEmpty()) {
                line = linesAfter.remove(0);
            } else if (it.hasNext()) {
                line = it.next();
            } else {
                break;
            }
            if (linesAfterCount > 0) {
                while (linesAfter.size() < linesAfterCount && it.hasNext()) {
                    linesAfter.add(it.next());
                }
            }

            Set<Pattern> searchPatterns = settings.getSearchPatterns()
                    .stream().filter(p -> !matchedPatterns.contains(p))
                    .collect(Collectors.toSet());
            for (Pattern p : searchPatterns) {
                Matcher m = p.matcher(line);
                boolean found = m.find();
                while (found) {
                    if ((linesBefore.isEmpty() || linesBeforeMatch(linesBefore))
                            &&
                            (linesAfter.isEmpty() || linesAfterMatch(linesAfter))
                            &&
                            (!settings.hasLinesAfterToOrUntilPatterns() ||
                                    linesAfterToOrUntilMatch(it, linesAfter))
                            )
                    {
                        SearchResult searchResult = new SearchResult(
                                p,
                                null,
                                lineNum,
                                m.start() + 1,
                                m.end() + 1,
                                line,
                                new ArrayList<>(linesBefore),
                                new ArrayList<>(linesAfter));
                        results.add(searchResult);
                        if (settings.getFirstMatch()) {
                            matchedPatterns.add(p);
                            found = false;
                        } else {
                            found = m.find(m.end());
                        }
                    } else {
                        found = false;
                    }
                }
            }

            if (linesBeforeCount > 0) {
                if (linesBefore.size() == linesBeforeCount) {
                    linesBefore.remove(0);
                }
                if (linesBefore.size() < linesBeforeCount) {
                    linesBefore.add(line);
                }
            }
            if (settings.getFirstMatch() &&
                    matchedPatterns.size() == settings.getSearchPatterns().size()) {
                break;
            }
        }
        return results;
    }

    private List<SearchResult> searchBinaryFile(final FileResult fr) {
        if (settings.getVerbose()) {
            log(String.format("Searching binary file %s", fr.path()));
        }
        List<SearchResult> results = new ArrayList<>();
        try {
            String content = FileUtil.getFileContents(fr.path(), StandardCharsets.ISO_8859_1);
            for (Pattern p : settings.getSearchPatterns()) {
                Matcher m = p.matcher(content);
                boolean found = m.find();
                while (found) {
                    results.add(new SearchResult(
                            p,
                            fr,
                            0,
                            m.start() + 1,
                            m.end() + 1,
                            ""));
                    found = !settings.getFirstMatch() && m.find(m.end());
                }
            }
        } catch (IOException | NoSuchElementException | IllegalStateException e) {
            log(e.toString());
        }
        return results;
    }

    public final void sortSearchResults(List<SearchResult> searchResults) {
        if (settings.getSortBy().equals(SortBy.FILENAME)) {
            searchResults.sort((sr1, sr2) -> sr1.compareByName(sr2, settings.getSortCaseInsensitive()));
        } else if (settings.getSortBy().equals(SortBy.FILESIZE)) {
            searchResults.sort((sr1, sr2) -> sr1.compareBySize(sr2, settings.getSortCaseInsensitive()));
        } else if (settings.getSortBy().equals(SortBy.FILETYPE)) {
            searchResults.sort((sr1, sr2) -> sr1.compareByType(sr2, settings.getSortCaseInsensitive()));
        } else if (settings.getSortBy().equals(SortBy.LASTMOD)) {
            searchResults.sort((sr1, sr2) -> sr1.compareByLastMod(sr2, settings.getSortCaseInsensitive()));
        } else {
            searchResults.sort((sr1, sr2) -> sr1.compareByPath(sr2, settings.getSortCaseInsensitive()));
        }
        if (settings.getSortDescending()) {
            Collections.reverse(searchResults);
        }
    }
}

File: kotlin/ktsearch/src/main/kotlin/ktsearch/Searcher.kt
package ktsearch

import ktfind.*
import java.io.IOException
import java.nio.charset.Charset
import kotlin.streams.toList

/**
 * @author cary on 7/23/16.
 */
class Searcher(val settings: SearchSettings) {
    private val finder: Finder
    private var charset: Charset? = null

    init {
        try {
            finder = Finder(settings.findSettings)
        } catch (e: FindException) {
            throw SearchException(e.message!!)
        }
        validateSettings(settings)
    }

    private fun validateSettings(settings: SearchSettings) {
        if (settings.searchPatterns.isEmpty()) {
            throw SearchException("No search patterns defined")
        }
        if (settings.linesAfter < 0) {
            throw SearchException("Invalid linesafter")
        }
        if (settings.linesBefore < 0) {
            throw SearchException("Invalid linesbefore")
        }
        if (settings.maxLineLength < 0) {
            throw SearchException("Invalid maxlinelength")
        }
        try {
            charset = Charset.forName(settings.textFileEncoding)
        } catch (_: IllegalArgumentException) {
            throw SearchException("Invalid or unsupported encoding: ${settings.textFileEncoding}")
        }
    }

    private fun anyMatchesAnyPattern(sList: List<String>,
                                     patternSet: Set<Regex>): Boolean {
        return sList.any { s -> matchesAnyPattern(s, patternSet) }
    }

    private fun matchesAnyPattern(s: String, patternSet: Set<Regex>): Boolean {
        return patternSet.any { p -> p.containsMatchIn(s) }
    }

    private fun findAnyMatches(s: String, startIndex: Int, patternSet: Set<Regex>): List<MatchResult> {
        return patternSet.mapNotNull { p -> p.find(s, startIndex) }
    }

    fun search(): List<SearchResult> {
        val fileResults = finder.find()
        if (settings.verbose) {
            val dirResults: List<String> = fileResults.asSequence()
                .map { it.path.parent }
                .map { it?.toString() ?: "." }
                .distinct().sorted().toList()
            log("\nDirectories to be searched (${dirResults.size}):")
            for (d in dirResults) {
                log(d)
            }
            log("\n\nFiles to be searched (${fileResults.size}):")
            for (fr in fileResults) {
                log(fr.toString())
            }
            log("")
        }

        return searchFiles(fileResults)
    }

    private fun searchFiles(frs: List<FileResult>): List<SearchResult> {
        val results = frs.flatMap { searchFile(it) }
        return sortSearchResults(results)
    }

    private fun searchFile(fr: FileResult): List<SearchResult> {
        return when {
            setOf(FileType.TEXT, FileType.CODE, FileType.XML).contains(fr.fileType) -> {
                searchTextFile(fr)
            }
            fr.fileType === FileType.BINARY -> {
                searchBinaryFile(fr)
            }
            else -> {
                listOf()
            }
        }
    }

    private fun searchTextFile(fr: FileResult): List<SearchResult> {
        if (settings.debug) {
            log("Searching text file $fr")
        }
        return if (settings.multiLineSearch) {
            searchTextFileContents(fr)
        } else {
            searchTextFileLines(fr)
        }
    }

    private fun searchTextFileContents(fr: FileResult): List<SearchResult> {
        val results: List<SearchResult> =
                searchMultilineString(fr.path.toFile().readText(charset!!))
        return results.map { r -> r.copy(file = fr) }
    }

    private fun getLinesFromMultiLineString(s: String, endLineIndices: List<Int>): List<String> {
        fun recGetLinesFromMultiLineString(indices: List<Int>, lines: List<String>): List<String> {
            return if (indices.size < 2) {
                lines
            } else {
                val nextLine = s.substring(indices.first() + 1, indices[1])
                recGetLinesFromMultiLineString(indices.drop(1), lines.plus(nextLine))
            }
        }
        return recGetLinesFromMultiLineString(endLineIndices, listOf())
    }

    private fun getLinesAfterFromMultiLineString(s: String,
                                                 afterNewlineIndices: List<Int>): List<String> {
        return if (afterNewlineIndices.isNotEmpty()) {
            val startIndex = afterNewlineIndices.first()
            if (settings.linesAfterToPatterns.isNotEmpty()) {
                val matches = findAnyMatches(s, startIndex,
                        settings.linesAfterToPatterns)
                if (matches.isNotEmpty()) {
                    val firstMatch = matches.minByOrNull { it.range.first }
                    val count = afterNewlineIndices.count { it <= firstMatch!!.range.first }
                    getLinesFromMultiLineString(s, afterNewlineIndices.take(count + 1))
                } else listOf()
            } else if (settings.linesAfterUntilPatterns.isNotEmpty()) {
                val matches = findAnyMatches(s, startIndex,
                        settings.linesAfterUntilPatterns)
                if (matches.isNotEmpty()) {
                    val firstMatch = matches.minByOrNull { it.range.first }
                    val count = afterNewlineIndices.count { it <= firstMatch!!.range.first }
                    getLinesFromMultiLineString(s, afterNewlineIndices.take(count))
                } else listOf()
            } else if (settings.linesAfter > 0) {
                val laIndices = afterNewlineIndices.take(settings.linesAfter + 1)
                getLinesFromMultiLineString(s, laIndices)
            } else listOf()
        } else listOf()
    }

    fun searchMultilineString(s: String): List<SearchResult> {
        val results: MutableList<SearchResult> = mutableListOf()
        val newlineIndices: List<Int> = s.indices.zip(s.asIterable()).
                filter { it.second == '\n' }.map { it.first }

        for (p in settings.searchPatterns) {
            val matches: Sequence<MatchResult> =
                    if (settings.firstMatch) {
                        val m = p.find(s)
                        if (m == null) {
                            sequenceOf()
                        } else {
                            sequenceOf(m)
                        }
                    }
                    else p.findAll(s)

            for (m in matches) {
                val beforeNewlineIndices = newlineIndices.takeWhile { it <= m.range.first }
                // val beforeLineCount = newlineIndices.count { it <= m.range.first }
                val linesBefore =
                        if (settings.linesBefore > 0) {
                            val lbIndices = beforeNewlineIndices.reversed().
                                    take(settings.linesBefore + 1).reversed()
                            getLinesFromMultiLineString(s, lbIndices)
                        } else listOf()
                val afterNewlineIndices = newlineIndices.dropWhile { it <= m.range.first }
                val linesAfter = getLinesAfterFromMultiLineString(s, afterNewlineIndices)
                if (!linesBeforeMatch(linesBefore) || !linesAfterMatch(linesAfter)) {
                    continue
                }
                val lineNum = beforeNewlineIndices.size + 1
                val startLineIndex =
                        if (lineNum == 1) 0
                        else beforeNewlineIndices.last() + 1
                val endLineIndex =
                        if (afterNewlineIndices.isNotEmpty()) afterNewlineIndices.first()
                        else s.length - 1
                val line = s.substring(startLineIndex, endLineIndex)
                val startMatchIndex = m.range.first - startLineIndex + 1
                val endMatchIndex = m.range.last - startLineIndex + 2
                results.add(SearchResult(
                        p,
                        null,
                        lineNum,
                        startMatchIndex,
                        endMatchIndex,
                        line,
                        linesBefore,
                        linesAfter))
            }
        }
        return results.toList()
    }

    fun searchTextFileLines(sf: FileResult): List<SearchResult> {
        val results: List<SearchResult> = sf.path.toFile().reader(charset!!).
                useLines { ss -> searchLineIterator(ss.iterator()) }
        return results.map { r -> r.copy(file = sf) }
    }

    private fun linesMatch(lines: List<String>, inPatterns: Set<Regex>,
                   outPatterns: Set<Regex>): Boolean {
        return (inPatterns.isEmpty() || anyMatchesAnyPattern(lines, inPatterns))
                &&
                (outPatterns.isEmpty() || !anyMatchesAnyPattern(lines, outPatterns))

    }

    private fun linesBeforeMatch(linesBefore: List<String>): Boolean {
        return linesBefore.isEmpty()
                ||
                linesMatch(linesBefore, settings.inLinesBeforePatterns,
                        settings.outLinesBeforePatterns)
    }

    private fun linesAfterMatch(linesAfter: List<String>): Boolean {
        return linesAfter.isEmpty()
                ||
                linesMatch(linesAfter, settings.inLinesAfterPatterns,
                        settings.outLinesAfterPatterns)
    }

    private fun matchLinesAfterToOrUntil(linesAfter: MutableList<String>,
                                         lines: Iterator<String>): Boolean {
        if (settings.linesAfterToPatterns.isEmpty()
                && settings.linesAfterUntilPatterns.isEmpty())
            return true

        for (i: Int in linesAfter.indices) {
            if (matchesAnyPattern(linesAfter[i], settings.linesAfterToPatterns)) {
                while (i + 1 < linesAfter.size) linesAfter.removeAt(i + 1)
                return true
            } else if (matchesAnyPattern(linesAfter[i], settings.linesAfterUntilPatterns)) {
                while (i < linesAfter.size) linesAfter.removeAt(i)
                return true
            }
        }
        var foundMatch = false
        while (!foundMatch && lines.hasNext()) {
            val nextLine = lines.next()
            when {
                matchesAnyPattern(nextLine, settings.linesAfterToPatterns) -> {
                    linesAfter.add(nextLine)
                    foundMatch = true
                }
                matchesAnyPattern(nextLine, settings.linesAfterUntilPatterns) -> {
                    foundMatch = true
                }
                else -> {
                    linesAfter.add(nextLine)
                }
            }
        }
        return foundMatch
    }

    fun searchLineIterator(lines: Iterator<String>): List<SearchResult> {
        val results: MutableList<SearchResult> = mutableListOf()
        val linesBefore: MutableList<String> = mutableListOf()
        val linesAfter: MutableList<String> = mutableListOf()
        var lineNum = 0
        val matchedPatterns : MutableSet<Regex> = mutableSetOf()
        while (true) {
            lineNum += 1
            val line =
                    if (linesAfter.isNotEmpty()) {
                        linesAfter.removeAt(0)
                    } else if (lines.hasNext()) {
                        lines.next()
                    } else {
                        break
                    }
            if (settings.linesAfter > 0) {
                while (linesAfter.size < settings.linesAfter && lines.hasNext()) {
                    linesAfter.add(lines.next())
                }
            }

            val searchPatterns =
                    if (settings.firstMatch)
                        settings.searchPatterns.filterNot { matchedPatterns.contains(it) }
                    else settings.searchPatterns

            if (searchPatterns.isEmpty()) {
                break
            }

            for (p in searchPatterns) {
                val matches: Sequence<MatchResult> =
                        if (settings.firstMatch) {
                            val m = p.find(line)
                            if (m == null) {
                                sequenceOf()
                            } else {
                                sequenceOf(m)
                            }
                        }
                        else p.findAll(line)
                if (matches.count() == 0
                        || !linesBeforeMatch(linesBefore)
                        || !linesAfterMatch(linesAfter)
                        || !matchLinesAfterToOrUntil(linesAfter, lines.iterator())) {
                    continue
                }

                matchedPatterns.add(p)

                for (m in matches) {
                    results.add(SearchResult(
                            p,
                            null,
                            lineNum,
                            m.range.first + 1,
                            m.range.last + 2,
                            line,
                            linesBefore.toList(),
                            linesAfter.toList()))
                }
            }
            if (settings.linesBefore > 0) {
                if (linesBefore.size == settings.linesBefore) {
                    linesBefore.removeAt(0)
                }
                if (linesBefore.size < settings.linesBefore) {
                    linesBefore.add(line)
                }
            }
            if (settings.firstMatch && matchedPatterns.size == settings.searchPatterns.size) {
                break
            }
        }
        return results.toList()
    }

    private fun searchBinaryFile(sf: FileResult): List<SearchResult> {
        if (settings.verbose) {
            log("Searching binary file $sf")
        }
        val results: MutableList<SearchResult> = mutableListOf()

        try {
            val content: String = sf.path.toFile().readText(Charsets.ISO_8859_1)
            for (p in settings.searchPatterns) {
                val matches: Sequence<MatchResult> =
                        if (settings.firstMatch) {
                            val m = p.find(content)
                            if (m == null) {
                                sequenceOf()
                            } else {
                                sequenceOf(m)
                            }
                        }
                        else p.findAll(content)
                matches.forEach { m ->
                    results.add(SearchResult(
                            p,
                            sf,
                            0,
                            m.range.first + 1,
                            m.range.last + 1,
                            ""))
                }
            }
        } catch (e: IOException) {
            log(e.toString())
        } catch (e: NoSuchElementException) {
            log(e.toString())
        } catch (e: IllegalStateException) {
            log(e.toString())
        }
        return results.toList()
    }

    private fun sortSearchResults(searchResults: List<SearchResult>): List<SearchResult> {
        val sortedSearchResults: MutableList<SearchResult> =
            when (settings.sortBy) {
                SortBy.FILENAME -> {
                    searchResults.stream()
                        .sorted { sr1, sr2 -> sr1.compareByName(sr2, settings.sortCaseInsensitive) }.toList()
                }
                SortBy.FILESIZE -> {
                    searchResults.stream()
                        .sorted { sr1, sr2 -> sr1.compareBySize(sr2, settings.sortCaseInsensitive) }.toList()
                }
                SortBy.FILETYPE -> {
                    searchResults.stream()
                        .sorted { sr1, sr2 -> sr1.compareByType(sr2, settings.sortCaseInsensitive) }.toList()
                }
                SortBy.LASTMOD -> {
                    searchResults.stream()
                        .sorted { sr1, sr2 -> sr1.compareByLastMod(sr2, settings.sortCaseInsensitive) }.toList()
                }
                else -> {
                    searchResults.stream()
                        .sorted { sr1, sr2 -> sr1.compareByPath(sr2, settings.sortCaseInsensitive) }.toList()
                }
            }.toMutableList()

        if (settings.sortDescending) {
            sortedSearchResults.reverse()
        }

        return sortedSearchResults.toList()
    }
}
